experiments/preprocessing/merge_and_handle_NaNs.py

```diff
@@ -0,0 +1,81 @@
+import pandas as pd
+
+def load_dataset(df_name):
+    if not df_name.endswith('.csv'):
+        raise ValueError("Only CSV files are supported.")
+    try:
+        df = pd.read_csv(df_name)
+        if df.empty or len(df.columns) == 1:
+            raise ValueError(f"Error in file '{df_name}' - empty or incorrect separator.")
+        df.columns = df.columns.str.lower()
+        return df
+    except (FileNotFoundError, pd.errors.EmptyDataError, pd.errors.ParserError, UnicodeDecodeError) as e:
+        raise ValueError(f"Error loading file '{df_name}': {e}")
+
+def fill_missing_values_shop_sales(df):
+    df['date_id'] = pd.to_numeric(df['date_id'], errors='coerce')
+    df['cnt'] = pd.to_numeric(df['cnt'], errors='coerce')
+    df[['item_id', 'store_id']] = df[['item_id', 'store_id']].astype(str)
+
+    max_date_id = df['date_id'].max()
+    for i in range(1, len(df)):
+        if pd.isna(df.at[i, 'date_id']):
+            if pd.isna(df.at[i - 1, 'date_id']):
+                raise ValueError(f"Consecutive NaN in 'date_id' at index {i}")
+            df.at[i, 'date_id'] = 1 if df.at[i - 1, 'date_id'] == max_date_id else df.at[i - 1, 'date_id'] + 1
+
+    df['item_id'].fillna(method='ffill', inplace=True)
+    df['store_id'] = df['item_id'].str.split('_').str[0]
+
+    return df
+
+def preprocess_shop_sales_dates(df):
+    df = df.drop(columns=['weekday']).reset_index(drop=True)
+    df[['wm_yr_wk', 'wday', 'month', 'year']] = df[['wm_yr_wk', 'wday', 'month', 'year']].apply(pd.to_numeric, errors='coerce')
+    df['date'] = pd.to_datetime(df['date'], errors='coerce')
+    
+    df['date_id'] = df['date_id'].interpolate().fillna(method='ffill')
+    df['date'] = df['date'].interpolate().fillna(method='ffill')
+    df['wday'], df['month'], df['year'] = df['date'].dt.weekday + 1, df['date'].dt.month, df['date'].dt.year
+
+    df[['event_name_1', 'event_type_1', 'event_name_2', 'event_type_2']] = df[['event_name_1', 'event_type_1', 'event_name_2', 'event_type_2']].fillna('Unknown')
+    df[['cashback_store_1', 'cashback_store_2', 'cashback_store_3']] = df[['cashback_store_1', 'cashback_store_2', 'cashback_store_3']].fillna(0)
+
+    return df
+
+def preprocess_shop_sales_prices(df):
+    df[['wm_yr_wk', 'sell_price']] = df[['wm_yr_wk', 'sell_price']].apply(pd.to_numeric, errors='coerce')
+    df['wm_yr_wk'] = df['wm_yr_wk'].interpolate().round().fillna(method='ffill')
+    df.sort_values(by=['item_id', 'wm_yr_wk'], inplace=True)
+    df['sell_price'] = df['sell_price'].interpolate()
+    return df
+
+def safe_merge(shop_sales, shop_sales_dates, shop_sales_prices):
+    for df in [shop_sales, shop_sales_dates, shop_sales_prices]:
+        df.columns = df.columns.str.lower()
+
+    merged_df = pd.merge(shop_sales, shop_sales_dates, on='date_id', how='outer')
```

Comment:
You are using the 'outer' method, but I don't see you checking the merged DataFrame for the presence of NaN values afterward. 

The 'outer' method retains all rows from both DataFrames, filling in NaN for missing matches from either side

So, in short, the 'outer' method can result in NaN values when there are no matches, which is the key difference between 'outer' and 'left' joins